src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/CpsPackageReferenceProject.cs

```diff
@@ -371,6 +371,56 @@ public override Task<string> GetCacheFilePathAsync()
             return Task.FromResult(NoOpRestoreUtilities.GetProjectCacheFilePath(cacheRoot: spec.RestoreMetadata.OutputPath));
         }
 
+        // To avoid race condition, we work on copy of cache.
+        protected override (List<FrameworkInstalledPackages> installedPackagesCopy, List<FrameworkInstalledPackages> transitivePackagesCopy) GetCacheCopy()
+        {
+            return (new List<FrameworkInstalledPackages>(InstalledPackages), new List<FrameworkInstalledPackages>(TransitivePackages));
+        }
+
+        protected override void UpdatePackageListDetails(List<FrameworkInstalledPackages> packages, IEnumerable<FrameworkInstalledPackages> detectedNewPackages)
+        {
+            if (packages == null || detectedNewPackages == null)
+            {
+                return;
+            }
+
+            var dict = new Dictionary<NuGetFramework, FrameworkInstalledPackages>();
```

Comment:
Can we rename dict to add more context?

```suggestion
            var packagesCopy = new Dictionary<NuGetFramework, FrameworkInstalledPackages>();
```

Reply:
I think we'd even need a better name than that. What packages copy? :D